alacritty/src/display/damage.rs

```diff
@@ -0,0 +1,345 @@
+use std::cmp;
+
+use glutin::Rect;
+
+use alacritty_terminal::selection::SelectionRange;
+use alacritty_terminal::term::{DamagedLine, RenderableCursor, SizeInfo, TermDamage};
+use alacritty_terminal::vi_mode::ViModeCursor;
+
+/// Minimum part of damaged square when intersecting damaged rectangles.
+pub const MIN_DAMAGED_SQUARE_PART: f32 = 0.4;
+
+/// An error when we for look ahead in iterator, used to try to collect rectangles which pretty
+/// close to each other, but don't fit into `MIN_DAMAGED_SQUARE_PART` just by looking on the first
+/// next damaged rectangle.
+pub const MIN_DAMAGED_SQUARE_PART_ERR: f32 = MIN_DAMAGED_SQUARE_PART / 2.;
+
+pub struct RendererDamage<'a> {
+    index: usize,
+    rect: Option<Rect>,
+    size_info: SizeInfo<u32>,
+    terminal_damage: &'a mut Vec<DamagedLine>,
+}
+
+impl<'a> RendererDamage<'a> {
+    pub fn new(
+        terminal_damage: &'a mut TermDamage,
+        size_info: SizeInfo<u32>,
+        cursor: RenderableCursor,
+        vi_cursor: Option<ViModeCursor>,
+        selection: Option<SelectionRange>,
+        display_offset: usize,
+    ) -> Self {
+        let fully_damaged = terminal_damage.take_fully_damaged();
+        let terminal_damage =
+            terminal_damage.damage_mut(cursor, vi_cursor, selection, display_offset);
+
+        let (index, rect) = if fully_damaged {
+            let rect = Rect { x: 0, y: 0, width: size_info.width(), height: size_info.height() };
+            (terminal_damage.len(), rect)
+        } else {
+            // Terminal cursor is always damaged there's at least one damaged line.
+            let index = terminal_damage.iter().position(|&line| line.is_damaged()).unwrap();
+            let rect = Self::rect_for_line(&size_info, index, terminal_damage[index]);
+
+            terminal_damage[index].reset();
+
+            // Advance index into the next unprocessed element.
+            (index + 1, rect)
+        };
+
+        Self { index, rect: Some(rect), terminal_damage, size_info }
+    }
+
+    /// Creates a new rectangle in a way that 'lhs' and 'rhs' are inside of it.
+    #[inline]
+    fn intersect(lhs: Rect, rhs: Rect) -> Rect {
```

Comment:
Intersection would be a rectangle containing only area both `lhs` and `rhs` contain. Your comment doesn't describe intersection, it describes a union (since it's rectangles it's even more than a true union).

The code also suggests this isn't an intersection, so please rename the function. I think smithay calls this `merge` for example(https://github.com/Smithay/smithay/blob/master/src/utils/geometry.rs#L955)?